### PR TITLE
MapView in-progress

### PR DESCRIPTION
**Pull request by Runi-VN**
Made some changes as to not overwrite the `user`-state and appended my info on it.  
Made `distance`-state in `App` as to hold ssot.

## NECESSARY

- [x] bug with settings & setting Distance.  the useEffect overwrites any changes.
lifted state up
- [ ] errorhandling the way @MalteMagnussen designed it. **no time**
- [x] when ready, convert test-"settings" to actually work with passed userprops
  - [x] Settings are used throughout the screen + component, so there are quite a few changes to be made.
- [x] Implement the facade
Most is done, but it haven't been tested yet. 
**Decision to be made - when to**: 
a) ~~update own location in backend~~ this is done already by fetching users
b) fetch nearby users
currently it is set to do it following the location update (1000ms) - test this, it may be too aggressive.

## Style/Code setup
- [ ] Investigate if mapscreenSettings lives up to design requirements (state)
  - [ ] mapscreensettings might have a view or two too much
  - [ ] button styling - i had the idea to make them transparent, ontop of the mapview. hm.
  - [ ] could maybe clean up the mapscreensettings prop in general. suggestions welcome.
  - [ ] Style the setting modal
- [ ] clean both stylesheets (duplicates, outcommented)

## Current looks:
1. Zooms to user location on startup (with given location permissions)
2. Zooms to user on request (onLongPress / platform-dependant button) 
image
3. updates location every 1000ms, only in foreground

image

<details><summary>Old</summary>
<p>
## Issues
- we cant use the MapView prop `showsUserLocation` as we can't annotate it (as far as I know).  
Disabling this also removes a lot of other features, such as 'zoom to user location`.
- having some issues setting region after first <ins>and only after first</ins> location load

## Progress
- [x] Dependencies:
`expo install expo-location`
`expo install expo-constants`
`expo install react-native-maps`
- [x] References:
  - https://docs.expo.io/versions/latest/sdk/map-view/
  - https://github.com/react-native-community/react-native-maps (documentation, examples)
  - https://docs.expo.io/versions/v37.0.0/sdk/location/
  - lat/lon delta:  https://stackoverflow.com/a/36688156
- [x] Set updateLocation to 1000ms
  - bug
The idea is for it to continue updating in the background, but if you swap back and forth between screens, we just initiate more and more TimeOuts, resulting in increasingly large amount of updateLocations...
   - [x] Investigate [taskmanager](https://docs.expo.io/versions/v37.0.0/sdk/task-manager/#example)
- [x] Investigate useRef. Compare to teachers implementation of mapview
- [x] Update wonky region setting.
  - [x] investigate animation events from initialRegion -> region
- [x] Prepare as much as possible for positionFacade
- [x] see if we can hide userLocation but still use its features. or just place a marker on top. :^) 

- [x] Investigate if we can use the Card component in the modal

**modal:**
~~modal 'meters' can't figure out how to display next to input as `[input] meters` or `1000 meters`~~  
~~modal keyboard dismiss dont work~~ either ignore or convert to scrollview
can't close modal by clicking outside it  
https://stackoverflow.com/questions/40483034/close-react-native-modal-by-clicking-on-overlay
</p>
</details>

### CODE REVIEW

**Comment by MalteMagnussen** [comment]
> The idea is for it to continue updating in the background, but if you swap back and forth between screens, we just initiate more and more TimeOuts, resulting in increasingly large amount of updateLocations...

This sounds to me like a clean-up issue. 
This kinda issue is exactly what the "Clean-up" function of a useEffect is for! :) 

Remember you can return something from a useEffect, and it will act as the Clean-Up. Specifically designed for things like this! Let me see if I can find an example. 

https://stackoverflow.com/a/53090848/11255140

Hope this was helpful, and that I understood the issue correctly.

**Comment by MalteMagnussen** [comment]
Discussion of different options for animating map to a position 
https://stackoverflow.com/questions/56469464/pros-cons-of-using-animatetoregion-vs-animatecamera

**Comment by MalteMagnussen** [comment]
>  Investigate useRef. Compare to teachers implementation of mapview

I found the official docs here. But I don't understand it. 
https://reactjs.org/docs/hooks-reference.html#useref

Teacher implementation: 
https://github.com/MalteMagnussen/SemestercaseMobileApp/blob/master/App.js#L135
Is set here
https://github.com/MalteMagnussen/SemestercaseMobileApp/blob/master/App.js#L69
And then called in useEffect upon region change? 
https://github.com/MalteMagnussen/SemestercaseMobileApp/blob/master/App.js#L73
I dont know. Hope it helps.

File: FindYourFriends/screens/MapScreen.js [hunk 1]
```diff
@@ -1,56 +1,178 @@
...
+				setSettings({
+					...settings,
+					longitude: location.coords.longitude,
+					latitude: location.coords.latitude,
+				});
+				setChangeRegion(true);
+				console.log('happened ' + new Date(Date.now()).toLocaleTimeString());
+			})();
+			//No return as we want user to update in the background
+		}, 1000);
```
Comment (MalteMagnussen) [inline FindYourFriends/screens/MapScreen.js]: This, I assume, needs a clean-up function :)

File: FindYourFriends/screens/MapScreen.js [hunk 2]
```diff
@@ -1,56 +1,178 @@
...
+				longitudeDelta: LONGITUDE_DELTA,
+			});
+		}
+	}, [changeRegion]);
+
+	useEffect(() => {
+		if (region) mapRef.current.animateToRegion(region, 1000);
+	}, [region]);
+	useEffect(() => {
+		setTimeout(() => {
```
Comment (MalteMagnussen) [inline FindYourFriends/screens/MapScreen.js]:
```suggestion
		const timeout = setTimeout(() => {
```

File: FindYourFriends/screens/MapScreen.js [hunk 3]
```diff
@@ -1,56 +1,178 @@
...
+					...settings,
+					longitude: location.coords.longitude,
+					latitude: location.coords.latitude,
+				});
+				setChangeRegion(true);
+				console.log('happened ' + new Date(Date.now()).toLocaleTimeString());
+			})();
+			//No return as we want user to update in the background
+		}, 1000);
+	}, []); //will replicate between reloads of component, ffs
```
Comment (MalteMagnussen) [inline FindYourFriends/screens/MapScreen.js]:
```suggestion
	return () => {
              clearTimeout(timeout);
        }
	}, []); //will replicate between reloads of component, ffs
```

**Comment by MalteMagnussen** [comment]
Converting to Draft, since it has the DO NOT MERGE label.

File: FindYourFriends/screens/MapScreen.js (hunk 1 above)
Comment (Runi-VN) [reply FindYourFriends/screens/MapScreen.js]: yeah, but then there's no updating in the background, which is what we want. as written in the pr, this should probably be handled by taskmanager.

**Comment by Runi-VN** [comment]
Finally got something working.
- [x] animateToRegion on initial load (and only initial load)
Had to fix a weeeird bug by setting a timeout of **5ms**. ty react-native-maps.
- [x] Setup background location handler, which works way better than the standard one. [Docs](https://docs.expo.io/versions/v37.0.0/sdk/location/#locationstartlocationupdatesasynctaskname-options)
- [x] Cleanup + unused imports
- [x] Documentation
- Integration w/ facade
- [x] Integrate [navigation](https://github.com/Hold-Krykke/Security_and_FullstackJavascript_Exam/pull/44)
- [x] Verify numbers only in distance
- Center input in settings  

For now I have to focus on the report though.

**Comment by Runi-VN** [comment]
We can actually monitor in the background without a timeout, but it eats up all the memory.

**Review by Castau — APPROVED** [review]

**Review by MalteMagnussen — APPROVED** [review]

**Review by MrHardcode — APPROVED** [review]

**Review by MrHardcode — APPROVED** [review]

**Pull request merged**